File: backend/chat/app/core/document_service.py
from app.utils.pdf_utils import extract_text_from_pdf
from app.core.vectorstore import VectorStore
from datetime import datetime
import tiktoken
import os
from app.core.providers.provider_factory import get_embedding_provider

ai_provider = get_embedding_provider()


def chunk_text(text: str, max_tokens: int = 500):
    """Divide texto em partes menores com base na contagem aproximada de tokens"""
    chunks = []
    current = []
    tokens_count = 0
    encoder = tiktoken.get_encoding("cl100k_base")

    for line in text.split("\n"):
        line_tokens = len(encoder.encode(line))
        if tokens_count + line_tokens > max_tokens:
            chunks.append("\n".join(current))
            current = []
            tokens_count = 0
        current.append(line)
        tokens_count += line_tokens

    if current:
        chunks.append("\n".join(current))
    return chunks


def ingest_pdf(file_bytes: bytes, filename: str, user_id: str):
    """Extrai texto de PDF, gera embeddings e salva no vetor store"""

    # 1️⃣ Lê o PDF e divide em chunks
    text = extract_text_from_pdf(file_bytes)
    chunks = chunk_text(text)

    # 2️⃣ Define provedor via variável de ambiente
    provider_name = os.getenv("EMBEDDING_PROVIDER", "google")
    vector_store = VectorStore()

    # 3️⃣ Gera embeddings e salva
    for i, chunk in enumerate(chunks, start=1):
        embedding = ai_provider.embed_text(chunk)
        metadata = {
            "file_name": filename,
            "chunk_index": i,
            "timestamp": datetime.utcnow().isoformat(),
            "user_id": user_id
        }
        vector_store.add(chunk, metadata, embedding)

    return {
        "message": f"Documento '{filename}' processado com sucesso usando {provider_name.title()}.",
        "chunks_count": len(chunks),
        "user_id": user_id
    }

File: backend/chat-v2/src/services/DocumentService.js
import { extractTextFromPdf } from '../utils/pdfUtils.js';
import VectorStore from '../core/VectorStore.js';
import DocumentRepository from '../repositories/DocumentRepository.js';
import { getEmbeddingProvider } from '../core/providers/providerFactory.js';

const embeddingProvider = getEmbeddingProvider();
const documentRepository = new DocumentRepository();

// Token counting approximation for JavaScript
function countTokens(text) {
  // Approximation: ~4 characters per token (similar to tiktoken)
  return Math.ceil(text.length / 4);
}

function chunkText(text, maxTokens = 500) {
  const chunks = [];
  const lines = text.split('\n');
  let current = [];
  let tokensCount = 0;

  for (const line of lines) {
    const lineTokens = countTokens(line);
    if (tokensCount + lineTokens > maxTokens) {
      chunks.push(current.join('\n'));
      current = [];
      tokensCount = 0;
    }
    current.push(line);
    tokensCount += lineTokens;
  }

  if (current.length > 0) {
    chunks.push(current.join('\n'));
  }

  return chunks;
}

async function ingestPdf(fileBytes, filename, userId, chatId = null) {
  try {
    // 1. Ler o PDF e dividir em chunks
    const text = await extractTextFromPdf(fileBytes);
    const chunks = chunkText(text);

    // 1.5. Persistir documento e obter document_id
    const documentId = await documentRepository.createDocument(userId, filename);

    // 2. Definir provedor via variável de ambiente
    const providerName = (process.env.EMBEDDING_PROVIDER || 'google').toLowerCase();
    const vectorStore = new VectorStore();

    // 3. Gerar embeddings e salvar
    for (let i = 0; i < chunks.length; i++) {
      const chunk = chunks[i];
      const embedding = await embeddingProvider.embedText(chunk);
      const metadata = {
        file_name: filename,
        chunk_index: i + 1,
        timestamp: new Date().toISOString(),
        user_id: userId,
        chat_id: chatId,
        document_id: documentId,
      };
      await vectorStore.add(chunk, metadata, embedding);
    }

    return {
      message: `Documento '${filename}' processado com sucesso usando ${providerName.charAt(0).toUpperCase() + providerName.slice(1)}.`,
      chunks_count: chunks.length,
      user_id: userId,
      document_id: documentId,
    };
  } catch (error) {
    throw new Error(`Erro ao ingerir PDF: ${error.message}`);
  }
}

export { ingestPdf };
